File: sonnet-scala/src/eval/Frame.scala
package eval

import syntax.Node
import scala.collection.mutable

class EvaluationError(message: String) extends RuntimeException(message)

enum Modifiability {
    case Mutable
    case Immutable
}
class Variable(val name: String, val mod: Modifiability)

class Frame(val parent: Option[Frame]) {

    def this(parentFrame: Frame) = {
        this(Some(parentFrame))
    }

    def this() = {
        this(None)
    }

    private val bindings: mutable.Map[Variable, Node] = mutable.Map.empty[Variable, Node]

    def define(name: String, value: Node, mutable: Boolean): Unit = {
        if (bindings.keys.exists(_.name == name)) {
            throw new EvaluationError(s"Identifier '$name' is already bound in this scope.")
        }
        val mod = if mutable then Modifiability.Mutable else Modifiability.Immutable
        bindings.put(Variable(name, mod), value)
    }

    def assign(name: String, value: Node): Unit = {
        bindings.keys.find(_.name == name) match {
            case Some(v) =>
                if (v.mod == Modifiability.Immutable) {
                    throw new EvaluationError(s"Cannot reassign immutable variable: $name")
                }
                bindings.put(v, value)
            case None =>
                parent match {
                    case Some(p) => p.assign(name, value)
                    case None    => throw new EvaluationError(s"Unbound variable: $name")
                }
        }
    }

    def lookup(name: String): Option[Node] = {
        bindings.find { case (v, _) => v.name == name } match {
            case Some((_, value)) => Some(value)
            case None =>
                parent match {
                    case Some(p) => p.lookup(name)
                    case None    => None
                }
        }
    }

    def contains(name: String): Boolean = {
        bindings.keys.exists(_.name == name)
    }

    def toStringBuilder(sb: StringBuilder): StringBuilder = {
        bindings.foreach { case (v, value) =>
            sb.append(s"${v.name} (${v.mod}) : $value\n")
        }
        parent.foreach { p =>
            p.toStringBuilder(sb)
        }
        sb
    }

    override def toString: String = {
        val sb = new StringBuilder("--- Frame ---\n")
        toStringBuilder(sb).append("\n-------------").toString()
    }
}

File: archives/sonnet-java/src/eval/Frame.java
package eval;

import java.util.HashMap;
import java.util.Map;
import syntax.AST;

public class Frame {
	private final Frame parent;
	private final Map<String, AST> bindings = new HashMap<>();

	public Frame(Frame parent) {
		this.parent = parent;
	}

	public Frame() {
		this.parent = null;
	}

	public void define(String name, AST value) {
		bindings.put(name, value);
	}

	public void assign(String name, AST value) {
		if (this.bindings.containsKey(name)) {
			this.bindings.put(name, value);
		} else if (this.parent != null) {
			this.parent.assign(name, value);
		} else {
			throw new EvaluationError("Unbound variable: " + name);
		}
	}

	public AST lookup(String name) {
		if (bindings.containsKey(name)) {
			return bindings.get(name);
		}
		return parent != null ? parent.lookup(name) : null;
	}

	public boolean contains(String name) {
		if (bindings.containsKey(name)) {
			return true;
		} else {
			return false;
		}
	}

	public StringBuilder toStringBuilder(StringBuilder sb) {
		bindings.forEach((key, value) -> sb.append(key + " : " + value + "\n"));
		if (parent != null) {
			parent.toStringBuilder(sb);
		}
		return sb;
	}

	@Override
	public String toString() {
		return toStringBuilder(new StringBuilder("--- Frame ---\n")).append("\n-------------").toString();
	}
}
